diff --git a/lib/ipc_us.c b/lib/ipc_us.c
index 8690de2..8047b62 100644
--- a/lib/ipc_us.c
+++ b/lib/ipc_us.c
@@ -264,7 +264,13 @@ retry_recv:
 	if (result == -1) {
 		if (errno == EAGAIN &&
 		    (processed > 0 || timeout == -1)) {
-			goto retry_recv;
+			result = qb_ipc_us_ready(one_way, timeout,
+						 POLLIN);
+			if (result == 0 || result == -EAGAIN) {
+				goto retry_recv;
+			}
+
+			return result;
 		} else if (errno == ECONNRESET || errno == EPIPE) {
 			qb_util_perror(LOG_DEBUG,
 				       "recv(fd %d) converting to ENOTCONN",
@@ -316,7 +322,17 @@ retry_recv:
 	if (result == -1) {
 		if (errno == EAGAIN &&
 		    (processed > 0 || timeout == -1)) {
-			goto retry_recv;
+			/*
+			 * Don't spin too hard else we can consume too
+			 * much cpu.
+			 */
+			result = qb_ipc_us_ready(one_way,
+						 100,
+						 POLLIN);
+			if (result == 0 || result == -EAGAIN) {
+				goto retry_recv;
+			}
+			return result;
 		} else {
 			return -errno;
 		}
