From e771439c1c7e4bbe51e9bb79db1a7a9c1ad2875e Mon Sep 17 00:00:00 2001
From: Andrew Beekhof <andrew@beekhof.net>
Date: Tue, 23 Oct 2012 11:14:59 +1100
Subject: [PATCH] IPC: Pass the timeout to poll() if the recv function returns EAGAIN

---
 lib/ipcc.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/lib/ipcc.c b/lib/ipcc.c
index 0abb0e7..703a11a 100644
--- a/lib/ipcc.c
+++ b/lib/ipcc.c
@@ -225,6 +225,7 @@ qb_ipcc_recv(struct qb_ipcc_connection * c, void *msg_ptr,
 {
 	int32_t res = 0;
 	int32_t res2 = 0;
+	int32_t poll_ms = 0;
 	if (c == NULL) {
 		return -EINVAL;
 	}
@@ -234,8 +235,9 @@ qb_ipcc_recv(struct qb_ipcc_connection * c, void *msg_ptr,
 		struct qb_ipc_one_way *ow = _response_sock_one_way_get(c);
 
 		if (ow == NULL) return res;
+                if (res == -EAGAIN) poll_ms = ms_timeout;
 
-		res2 = qb_ipc_us_ready(ow, 0, POLLIN);
+		res2 = qb_ipc_us_ready(ow, poll_ms, POLLIN);
 		if (res2 < 0) {
 			res = res2;
 		}
-- 
1.7.1

